GAS-60: Alerting automation

theTibi opened the pull request:
* Adding all Alert templates from current automation to new
* Adding filters controled by pmm_alert_filters
* Adding alerts only for specific services controled by pmm_alert_filters
* Removing percona specific variable names

theTibi committed:
GAS-60: Alerting automation
* Adding all Alert templates from current automation to new
* Adding filters controled by pmm_alert_filters
* Adding alerts only for specific services controled by pmm_alert_filters
* Removing percona specific variable names

theTibi committed:
* moving percona specific variables into inventory

theTibi committed:
* changing back default

cezmunsta commented on roles/pmm/defaults/main.yaml:
```diff
@@ -22,9 +22,28 @@ pmm_agent_setup_flags:
+  med_interval: 5
+  high_interval: 10
+  max_interval: 15
+pmm_alert_filters: []            # example: '{"VmalertAlertsError":{"filters":[{"label":"env","regexp":"staging","type":"MATCH"},{"label":"node_name","regexp":"db10","type":"MISMATCH"}]}}'
```
@theTibi you have the variable defined as a list and then show a string. See [`pmm_manage_users`](https://github.com/cezmunsta-org/gascan-ansible/blob/b2c78bc258323497997074a3ed3c93b554d80e69/roles/pmm/defaults/main.yaml#L76) as an example

cezmunsta commented on roles/pmm/defaults/main.yaml:
```diff
@@ -22,9 +22,28 @@ pmm_agent_setup_flags:
+pmm_alert_filters: []            # example: '{"VmalertAlertsError":{"filters":[{"label":"env","regexp":"staging","type":"MATCH"},{"label":"node_name","regexp":"db10","type":"MISMATCH"}]}}'
 
+## Alert Manager Rules Thresholds
+pmm_alertmanager_thr_PostgresqlCommitRateLow: 80
```
This should be defined a different way, it means that there is going to need to be a value for anything with a threshold.
I don't see why these need to be present as the values would be set when creating the alert rule and so it doesn't matter what is set in the template.

theTibi replied:
No, this is set in the template not in the alert rule. In the alert rule you do not change the expression.

cezmunsta replied:
So, how are rules being created from a single template with 2 different thresholds?

theTibi replied:
that's two different template.  You can not edit the expression even on GUI when you are creating an alert.

theTibi replied:
Now I have added `params` to all the templates, thresholds can be overwritten now from inventory but the syntax is different for alert with thresholds and without them.  The API is not smart enough to use the default values from the template you have to redefine them  again when you create the alert.. anyhow now this feature is added. will push it soon.

cezmunsta commented on roles/pmm/tasks/alerting-alerts.yaml:
```diff
@@ -1,3 +1,11 @@
+
+- name: Set Alert Filter
+  ansible.builtin.set_fact:
+    pmm_alerting_filters: "{% if pmm_alert_item.name is defined and pmm_alert_item.name in pmm_alert_filters %}{{ pmm_alert_filters[pmm_alert_item.name][\"filters\"] | default([])}}{% else %}[]{% endif %}"
```
What is this? This isn't how other conditional values are being set.

theTibi replied:
For every alert rule it has to check the filters if the alert rule name is in there, if not it need to have a default. This is the only way how I was able to achieve that but if you have better ideas please let me know.

cezmunsta replied:
You seem to be missing the point, go and look how the other conditional values are being set.. certainly not like this

cezmunsta replied:
For example: 
```
roles/pmm/vars/main.yaml
77:   engine: '{{ "DISCOVER_RDS_POSTGRESQL" if group_names | intersect(["postgresql"]) else "DISCOVER_RDS_MYSQL" }}'
80:   port: '{{ 5432 if group_names | intersect(["postgresql"]) else 3306 }}'
```

cezmunsta commented on roles/pmm/vars/main.yaml:
```diff
@@ -27,6 +27,8 @@ pmm_grafana_alerting_notification: '{{ pmm_grafana_api }}/alertmanager/grafana/c
 
 pmm_setup_alerting: true
 notification_services: ["mysql","mongo","generic","agent","proxysql","postgresql","backup"]
+# Available alert services ["mysql","mongo","generic","agent","proxysql","postgresql","backup"]
```
This requires a commented-out example. See [pmm_manage_users](https://github.com/cezmunsta-org/gascan-ansible/blob/b2c78bc258323497997074a3ed3c93b554d80e69/roles/pmm/defaults/main.yaml#L76) as an example

theTibi replied:
This got removed as I reworked this part.

cezmunsta commented:
https://github.com/cezmunsta-org/gascan-ansible/pull/2/files#diff-5c2b1d975c1f80e9a71539272084067ea7d6f2e1793b8d31ba0a3dad507eafa0R17-L18 seems to be missing configuration options too

theTibi committed:
* Updates templates if they are already exist
* Adding custom lables
* Adding filters
* Adding threshold
* Able to create multiply alerts from the same template with different filters, thresholds etc..
* Re-creates missing alert rules

cezmunsta commented on roles/pmm/vars/main.yaml:
```diff
@@ -24,11 +25,62 @@ pmm_grafana_alerting_rules_list: '{{ pmm_grafana_api }}/ruler/grafana/api/v1/rul
-# Available alert services ["mysql","mongo","generic","agent","proxysql","postgresql","backup"]
-pmm_enable_alerts_services: []
+
+pmm_alert_list: []
```
Why has this moved from defaults to vars and so is no longer usable? Also, why have you made it necessary to defined as JSON????

cezmunsta commented on roles/pmm/tasks/grafana-alerting.yaml:
```diff
@@ -71,10 +73,42 @@
   vars:
     pmm_alert_item: '{{ item }}'
-  loop: "{{ pmm_alerting_templates['templates'] | selectattr('labels.service','in', pmm_enable_alerts_services) }}"
+  loop: "{{ pmm_alert_list | from_json | selectattr('name','defined') | list }}"
```
Why has this been designed to require `from_json`?

cezmunsta commented on roles/pmm/tasks/alerting-templates.yaml:
```diff
@@ -19,3 +19,23 @@
+      - 409
+    body_format: json
+    validate_certs: '{{ pmm_server_insecure }}'
+  when: (pmm_template_item.name is defined and pmm_template_item.name in pmm_existing_alert_templates)
```
Newline at EOF is missing

theTibi committed:
* `pmm_alert_list` should be a json object

theTibi committed:
* add new line to end of file

cezmunsta commented on roles/pmm/defaults/main.yaml:
```diff
@@ -22,9 +22,79 @@ pmm_agent_setup_flags:
+pmm_alertmanager_thr_StaleUploadLog: 3900
+
+#pmm_alert_list: {}
+pmm_alert_list:
```
@theTibi this is still JSON

theTibi committed:
* pmm_alert_list moved to yaml

theTibi committed:
* introducing `pmm_enable_all_alerts_for_service` and `pmm_enable_all_templates_for_service` to control which alerts and template services do we want to enable
pmm_enable_all_alerts_for_service: ["mysql","generic"] , pmm_enable_all_templates_for_service: ["mysql","generic"] will enable all alert templates and alerts for mysql and generic. However with `pmm_alert_list` we can still override the individual alert parameters like filters,labels, thresholds etc..

theTibi committed:
* moving variables into defaults

cezmunsta committed:
Fixed non-standard variable name

theTibi committed:
* fixing node downs expression, they do not alert with the current rule

cezmunsta committed:
Merge branch 'main' into GAS-60
